cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/AuthBinding.java

```diff
@@ -614,71 +620,81 @@ private SentryGenericServiceClient getClient() throws Exception {
     return SentryGenericServiceClientFactory.create(authConf);
   }
 
-  /**
-   * Maps a {@link TAuthorizable sentry authorizable} to {@link EntityId}
-   *
-   * @param tAuthorizable the TAuthorizable which needs to be mapped to entity
-   * @param parent the parent entity of this TAuthorizable
-   * @return {@link EntityId} for the given {@link TAuthorizable}
-   */
-  private EntityId toEntityId(TAuthorizable tAuthorizable, @Nullable EntityId parent) {
-    Authorizable authorizable = ModelAuthorizables.from(tAuthorizable.getType(), tAuthorizable.getName());
+  private EntityType toEntityId(TAuthorizable tAuthorizable, Map<EntityType, String> entityParts) {
+    Authorizable sentryAuthorizable = ModelAuthorizables.from(tAuthorizable.getType(), tAuthorizable.getName());
     switch (Authorizable.AuthorizableType.valueOf(tAuthorizable.getType())) {
       case INSTANCE:
-        return new InstanceId(instanceName);
+        entityParts.put(EntityType.INSTANCE, instanceName);
+        return EntityType.INSTANCE;
       case NAMESPACE:
-        Namespace namespace = (Namespace) authorizable;
-        return new NamespaceId(namespace.getName());
+        Namespace namespace = (Namespace) sentryAuthorizable;
+        entityParts.put(EntityType.NAMESPACE, namespace.getName());
+        return EntityType.NAMESPACE;
       case ARTIFACT:
-        Artifact artifact = (Artifact) authorizable;
-        assertNonNullParent(parent, Authorizable.AuthorizableType.ARTIFACT);
-        // TODO: Ideally we should be constructing cdap authorizable from the tAuthorizable. CDAP authorizable does
-        // not have a concept of version so we pass a random version here just to create artifact id. The version will
-        // dropped automatically when an Authorizable will be created through it to list privilege.
-        return ((NamespaceId) parent).artifact(artifact.getName(), "1.0-SNAPSHOT");
+        Artifact artifact = (Artifact) sentryAuthorizable;
+        Preconditions.checkArgument(entityParts.containsKey(EntityType.NAMESPACE), "Artifact must belong to a " +
+          "namespace");
+        entityParts.put(EntityType.ARTIFACT, artifact.getName());
+        return EntityType.ARTIFACT;
       case APPLICATION:
-        Application application = (Application) authorizable;
-        assertNonNullParent(parent, Authorizable.AuthorizableType.APPLICATION);
-        return ((NamespaceId) parent).app(application.getName());
+        Application application = (Application) sentryAuthorizable;
+        Preconditions.checkArgument(entityParts.containsKey(EntityType.NAMESPACE), "Application must belong to a " +
```

Comment:
Also it would be good to print the full authorizable name here so that we can use it for debugging. Otherwise the only thing we'll see in logs is `Application must belong to a namespace`, and won't know how to debug.